sources/Map/LuaMap.cpp

```diff
@@ -0,0 +1,282 @@
+//
+//
+//
+
+#include <algorithm>
+#include <map>
+#include <set>
+#include "Map.hpp"
+#include "LuaMap.hpp"
+
+namespace BBM
+{
+	LuaMap::LuaMap()
+	: _map(17, std::vector<int>(17, 0)), _danger(17, std::vector<int>(17, 0)),
+	_player(), _roundedPlayer()
+	{
+	}
+
+	LuaMap::~LuaMap()
+	{ }
+
+	void LuaMap::clearDanger(void)
+	{
+		for (int i = 0; i < 17; i++) {
+			for (int j = 0; j < 17; j++) {
+				_danger[i][j] = 0;
+			}
+		}
+	}
+
+	bool LuaMap::setDanger(int xpos, int ypos, int dangerLevel)
+	{
+		if (xpos < 0 || xpos > 16 ||
+			ypos < 0 || ypos > 16)
+			return false;
+		if (_map[ypos][xpos] == MapGenerator::BREAKABLE ||
+			_map[ypos][xpos] == MapGenerator::UNBREAKABLE)
+			return false;
+		_danger[ypos][xpos] = dangerLevel;	
+		return true;
+	}
+
+	void LuaMap::setPlayer(Vector3f pos)
+	{
+		_player = Vector2f(pos.x, pos.z);
+		_roundedPlayer = Vector2f(std::round(_player.x), std::round(_player.y));
+	}
+
+	std::vector<Vector2f> LuaMap::fillPath(std::vector<Vector2f> &path, 
+	std::unordered_map<Vector2f, Vector2f> &cameFrom, Vector2f node) const
+	{
+		if (cameFrom.find(node) != cameFrom.end()) {
+			path.insert(path.begin(), cameFrom[node]);
+			return fillPath(path, cameFrom, cameFrom[node]);
+		} else
+			return path;
+	}
+
+	std::vector<Vector2f> LuaMap::getNeighbors(Vector2f node) const
+	{
+		std::vector<Vector2f> neighbors;
+		for (auto &dir : _dirs) {
+			Vector2f neighbor(node.x + dir.x, node.y + dir.y);
+			if (neighbor.y < 0 || neighbor.x < 0)
+				continue;
+			if (neighbor.y >= 17 || neighbor.x >= 17)
+				continue;
+			if (_map[neighbor.y][neighbor.x] == 0 &&
+				_danger[neighbor.y][neighbor.x] != 1)
+				neighbors.push_back(neighbor);
+		}
+		if (neighbors.size())
+			return neighbors;
+		for (auto &dir : _dirs) {
+			Vector2f neighbor(node.x + dir.x, node.y + dir.y);
+			if (neighbor.y < 0 || neighbor.x < 0)
+				continue;
+			if (neighbor.y >= 17 || neighbor.x >= 17)
+				continue;
+			if (_map[neighbor.y][neighbor.x] == 0)
+				neighbors.push_back(neighbor);
+		}
+		return neighbors;
+	}
+
+	std::vector<Vector2f> LuaMap::pathfind(Vector2f root, Vector2f target) const
+	{
+		std::vector<Vector2f> closed;
+		std::vector<Vector2f> open;
+		std::unordered_map<Vector2f, Vector2f> cameFrom;
+		std::unordered_map<Vector2f, double> gScore;
+		std::unordered_map<Vector2f, double> fScore;
+		std::vector<Vector2f> path;
+
+		open.push_back(root);
+		gScore[root] = 0;
+		fScore[root] = root.distance(target);
+
+		while (open.size())
+		{
+			auto min_elem = std::min_element(open.begin(), open.end(),
+			[&fScore](const decltype(open)::value_type &l, const decltype(open)::value_type &r) -> bool
```

Comment:
use const auto &, you don't need a declytype here.

Reply:
fixed -> 0d49984